tox.ini

```diff
@@ -87,21 +85,21 @@ description = Bump all test depeendencies
 basepython = python3
 deps =
   pre-commit>=2.6.0
-  pip-tools>=5.4.0
   setuptools>=51.1.1
 skip_install = true
+setenv =
+  # without his upgrade would likely not do anything
+  PIP_CONSTRAINT = /dev/null
 commands =
-  # --no-annotate is used in order to prevent conflicts with PRs produced by dependency-bot
-  pip-compile --upgrade --no-annotate --output-file=test-requirements.txt setup.py test-requirements.in
-  pip-compile --upgrade --no-annotate --output-file=docs/requirements.txt setup.py docs/requirements.in
-  {envpython} -m pre_commit run {posargs:--all-files --hook-stage manual -v}
+  # upgrade deps is just a pre-commit manual hook
+  {envpython} -m pre_commit run {posargs:--all-files --show-diff-on-failure
+ --hook-stage manual}
 
 [testenv:docs]
 description = Builds docs
 basepython = python3
 deps =
-  -r{toxinidir}/docs/requirements.in
-  -c{toxinidir}/docs/requirements.txt
+  -r{toxinidir}/docs/requirements.txt
```

Comment:
Having unpinned docs deps isn't the brightest idea: this hurts reproducibility/predictability.

Reply:
That is not removing it, we use PIP_CONSTRAINT env var to achieve the same locking, is defined inside root test env. I know is not obvious but is more reliable as it does apply everywhere including on pre-commit and on build stage when installing packages, something that `-c` is not able to achieve. Also there is another special case where I found -c not to be working: if you have an empty `deps =`  with only -c, tox will fail to run, but when using PIP_CONSTRAINT it will work (you can guess why).